// skillfish/dist/utils.js
/**
 * Utility functions for skillfish CLI.
 * These are pure functions extracted for testability.
 */
import { normalize, isAbsolute, basename } from 'path';
/**
 * Validates a path to prevent directory traversal attacks.
 * Ensures path doesn't escape the intended directory.
 */
export function isValidPath(pathStr) {
    // Reject absolute paths
    if (isAbsolute(pathStr))
        return false;
    // Normalize and check for directory traversal
    const normalized = normalize(pathStr);
    if (normalized.startsWith('..') || normalized.includes('/../'))
        return false;
    // Only allow alphanumeric, dots, hyphens, underscores, and forward slashes
    if (!/^[\w./-]+$/.test(pathStr))
        return false;
    // Reject paths that could be problematic
    if (pathStr.includes('//') || pathStr.startsWith('/'))
        return false;
    return true;
}
/**
 * Type guard for GitHub tree API response.
 * Validates the response structure at runtime.
 */
export function isGitTreeResponse(data) {
    if (typeof data !== 'object' || data === null)
        return false;
    const obj = data;
    // tree is optional, but if present must be an array
    if ('tree' in obj && obj.tree !== undefined) {
        if (!Array.isArray(obj.tree))
            return false;
        // Validate each item has required fields
        for (const item of obj.tree) {
            if (typeof item !== 'object' || item === null)
                return false;
            const entry = item;
            if (typeof entry.path !== 'string')
                return false;
            if (typeof entry.type !== 'string')
                return false;
        }
    }
    return true;
}
/**
 * Parse YAML frontmatter from SKILL.md content.
 * Extracts name and description fields with fallbacks.
 */
export function parseFrontmatter(content) {
    // Match frontmatter block: ---\n...\n---
    const match = content.match(/^---\r?\n([\s\S]*?)\r?\n---/);
    if (!match)
        return {};
    const yaml = match[1];
    // Extract name (handles quoted and unquoted values)
    const nameMatch = yaml.match(/^name:\s*["']?(.+?)["']?\s*$/m);
    const name = nameMatch?.[1]?.trim();
    // Extract description (handles quoted and unquoted values)
    const descMatch = yaml.match(/^description:\s*["']?(.+?)["']?\s*$/m);
    const description = descMatch?.[1]?.trim();
    return { name, description };
}
/**
 * Derive skill name from path and repo name.
 */
export function deriveSkillName(skillPath, repoName) {
    if (skillPath === 'SKILL.md' || skillPath === './SKILL.md') {
        return repoName;
    }
    const normalized = skillPath.replace(/\/SKILL\.md$/i, '');
    const name = basename(normalized);
    if (!/^[\w.-]+$/.test(name)) {
        return repoName;
    }
    return name;
}
/**
 * Convert kebab-case or snake_case to Title Case.
 * "skill-lookup" → "Skill Lookup"
 * "my_cool_skill" → "My Cool Skill"
 */
export function toTitleCase(str) {
    return str.replace(/[-_]/g, ' ').replace(/\b\w/g, (char) => char.toUpperCase());
}
/**
 * Truncate text to a maximum length, adding ellipsis if needed.
 */
export function truncate(text, maxLength) {
    if (text.length <= maxLength)
        return text;
    return text.slice(0, maxLength - 1).trim() + '…';
}
/**
 * Extract SKILL.md paths from validated GitHub tree response.
 */
export function extractSkillPaths(data, skillFilename = 'SKILL.md') {
    if (!data.tree)
        return [];
    return data.tree
        .filter((item) => item.type === 'blob' && item.path.endsWith(skillFilename))
        .map((item) => item.path);
}
/**
 * Sleep for a specified duration.
 */
export function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}
/**
 * Process items with bounded concurrency.
 * Prevents overwhelming resources with unbounded parallel requests.
 *
 * @param items - Items to process
 * @param fn - Async function to apply to each item
 * @param concurrency - Maximum concurrent operations (default: 10)
 */
export async function batchMap(items, fn, concurrency = 10) {
    const results = [];
    let index = 0;
    async function worker() {
        while (index < items.length) {
            const currentIndex = index++;
            results[currentIndex] = await fn(items[currentIndex]);
        }
    }
    // Start workers up to concurrency limit
    const workers = Array(Math.min(concurrency, items.length))
        .fill(null)
        .map(() => worker());
    await Promise.all(workers);
    return results;
}
/**
 * Create a fresh JSON output object for the add command.
 */
export function createJsonOutput() {
    return {
        success: true,
        installed: [],
        skipped: [],
        errors: [],
    };
}
/**
 * Create a fresh JSON output object for the submit command.
 */
export function createSubmitJsonOutput() {
    return {
        success: true,
        submitted: [],
        failed: [],
        skills_found: [],
        errors: [],
    };
}
/**
 * Check if stdout is a TTY (interactive terminal).
 */
export function isTTY() {
    return process.stdout.isTTY === true;
}
/**
 * Check if stdin is a TTY (interactive terminal).
 */
export function isInputTTY() {
    return process.stdin.isTTY === true;
}
